port image_rotate on ros2

yechun1 opened the pull request:
1. port image_rotate on ros2  
    * replace nodelet(nodelet::Nodelet) with class_loader(rclcpp::Node)
    * use ros2param instead of dymanic_reconfigure
    * update image_transport APIs
    * use tf2 time with header.stamp operation
    * replace other ros APIs with ROS2 APIs
    
    for test: verified based on ros2_intel_realsense package as camera image input source.
        `ros2 run image_rotate image_rotate image:=/camera/color/image_raw camera_info:=/camera/color/camera_info rotated/image:=/camera/color/image_raw_rotated`
        (or by launch) 
      `ros2 launch image_rotate image_rotate.launch.py`
    
    set param: by ros2param as rqt_reconfigure similar UI tool has not ported yet on ROS2
        `ros2 param set /ImageRotateNode target_x 0.5`

2. modify code to follow ros2 coding style

yechun1 committed:
port image_rotate on ros2

* replace nodelet(nodelet::Nodelet) with class_loader(rclcpp::Node)
* use ros2param instead of dymanic_reconfigure
* update image_transport APIs
* use tf2 time with header.stamp operation
* replace other ros APIs with ROS2 APIs

Test:
  1. Start camera, for example run ros2_intel_realsense as camera raw image input.
  2. run rviz2 to add two images with /camera/color/image_raw and /camera_color/image_raw_rotated
  3. ros2 run image_rotate image_rotate image:=/camera/color/image_raw camera_info:=/camera/color/camera_info rotated/image:=/camera/color/image_raw_rotated
    (or by launch) ros2 launch image_rotate image_rotate.launch.py
  4. Set param: by ros2param as rqt_reconfigure similar UI tool has not ported yet on ROS2
    ros2 param set /ImageRotateNode target_x 0.5

yechun1 committed:
modify code to follow ros2 coding style

yechun1 commented:
@mjcarroll image_rotate has been ported on ROS2, just update the APIs to ROS2 and the feature is keep the same with previous. I had verified based on ROS2 RealSense camera as raw image input. Would you please kindly help to review the code, many thanks.

mjcarroll commented:
Great, I'll put this on my list to review.

I think that you'll also be interested to know that we are working on some launch improvements for composable nodes, which should directly impact some of the `image_pipeline` work.

JWhitleyWork commented:
@yechun1 - Please rebase your branch on the `ros2` branch. This will bring in fixes from `melodic` and the new CI.

yechun1 commented:
thanks for enable the CI, finished code rebase, please help review.

yechun1 commented:
Test:
      1. Start camera, for example run ros2_intel_realsense as camera raw image input.
      2. run rviz2 to add two images with /camera/color/image_raw and /camera_color/image_raw_rotated
      3. ros2 run image_rotate image_rotate image:=/camera/color/image_raw camera_info:=/camera/color/camera_info rotated/image:=/camera/color/image_raw_rotated
        (or by launch) ros2 launch image_rotate image_rotate.launch.py
      4. Set param: by ros2param as rqt_reconfigure similar UI tool has not ported yet on ROS2
        ros2 param set /ImageRotateNode target_x 0.5

JWhitleyWork commented on image_rotate/CMakeLists.txt:
```diff
@@ -1,32 +1,64 @@
+rclcpp_register_node_plugins(${PROJECT_NAME} "${PROJECT_NAME}::ImageRotateNode")
+set(node_plugins "${node_plugins}${PROJECT_NAME}::ImageRotateNode;$<TARGET_FILE:ImageRotateNode>\n")
+
 install(TARGETS image_rotate
```
Could be replaced with ${PROJECT_NAME}

yechun1 replied:
yes, fixed by https://github.com/ros-perception/image_pipeline/pull/385/commits/45bd9d037c3a83e888b9844eb2c74e1fe22cce74

JWhitleyWork commented on image_rotate/CMakeLists.txt:
```diff
@@ -1,32 +1,64 @@
+        DESTINATION lib
 )
 
 add_executable(image_rotate_exe src/node/image_rotate.cpp)
```
I know it was called this previously, but I'm not a big fan of the _exe designation, especially now that ROS2 supports Windows. This would be confusing for a Windows user (with .exe being the executable extension). I would prefer something like _node or _bin or pretty much anything else.

yechun1 replied:
Me too. 
fixed by https://github.com/ros-perception/image_pipeline/pull/385/commits/45bd9d037c3a83e888b9844eb2c74e1fe22cce74

JWhitleyWork commented on image_rotate/CMakeLists.txt:
```diff
@@ -1,32 +1,64 @@
-install(FILES nodelet_plugins.xml
-        DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+install(TARGETS image_rotate_exe
+        DESTINATION lib/${PROJECT_NAME}
```
Why is this being installed twice?

yechun1 replied:
The `bin` installed is for running `image_rotate` directly and needn't run with `ros2 run`,  seems duplicated. Just removed on https://github.com/ros-perception/image_pipeline/pull/385/commits/45bd9d037c3a83e888b9844eb2c74e1fe22cce74

JWhitleyWork commented on image_rotate/CMakeLists.txt:
```diff
@@ -1,32 +1,64 @@
+        DESTINATION lib/${PROJECT_NAME}
 )
+install(DIRECTORY examples/launch
+  DESTINATION share/${PROJECT_NAME}/
```
This should probably be `share/${PROJECT_NAME}/launch/`.

yechun1 replied:
this is correct, the "ros2 launch xxx" work well.
Ref to https://github.com/ros2/demos/blob/d2a934bf0225aa1b468c164bc915e6a0b9286095/composition/CMakeLists.txt#L198

JWhitleyWork commented on image_rotate/include/image_rotate/image_rotate_nodelet.hpp:
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2008, Willow Garage, Inc.
```
Nodelet is probably not the correct designation for this anymore since there is no longer a concept of a "nodelet" in ROS2. I would just say `_node`.

yechun1 replied:
There will have filename changes and not easy for review. So just update in a separate commit, fixed on https://github.com/ros-perception/image_pipeline/pull/385/commits/727347be3a393c05d3cdb6cb483b93565d244722

JWhitleyWork commented on image_rotate/include/image_rotate/visibility.h:
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2008, Willow Garage, Inc.
```
Should be `.hpp`.

yechun1 replied:
Not sure if I could rename to .hpp directly, I have just copy-modify the file from https://github.com/ros2/demos/blob/master/composition/include/composition/visibility_control.h

JWhitleyWork commented on image_rotate/package.xml:
```diff
@@ -28,30 +28,27 @@
+  <build_depend>rclcpp</build_depend>
   <build_depend>tf2</build_depend>
   <build_depend>tf2_geometry_msgs</build_depend>
   <build_depend>tf2_ros</build_depend>
```
`<build_depend>` and `<exec_depend>` can be combined with `<depend>` in Package Format 2 (and 3).

yechun1 replied:
fixed

JWhitleyWork commented on image_rotate/src/node/image_rotate.cpp:
```diff
@@ -1,54 +1,54 @@
-*********************************************************************/
-#include <ros/ros.h>
-#include <nodelet/loader.h>
+// Copyright (c) 2008, Willow Garage, Inc.
```
Might be good to update at least the copyright year.

yechun1 replied:
I am not sure about the copyright, I would like to keep the original while porting. We could update it later.

JWhitleyWork reviewed (changes requested):
My changes are mostly semantic and some are subjective. Let me know once you've looked them over and I'll approve.

yechun1 committed:
update package.xml and cmakelist for readable

yechun1 committed:
rename filename and strings from nodelet to node

Nodelet is probably not the correct designation for this anymore since there is no longer a concept of a "nodelet" in ROS2, use "_node" instead.

Signed-off-by: Chris Ye <chris.ye@intel.com>

yechun1 commented:
@JWhitleyAStuff thanks for nice comments, I have updated the changes, would you please review again?

yechun1 commented:
Hope this package could be merged and publish on ROS2 D release :)
